nanoFramework.Tools.DebugLibrary.Shared/NFDevice/GlobalExclusiveDeviceAccess.cs

```diff
@@ -11,98 +11,294 @@ namespace nanoFramework.Tools.Debugger.NFDevice
+                $"{address.Host}:{address.Port}",
+                millisecondsTimeout,
+                cancellationToken);
         }
+
         #endregion
 
         #region Implementation
-        private static bool DoCommunicateWithDevice(string connectionKey, Action communication, int millisecondsTimeout, CancellationToken? cancellationToken)
+
+        private static GlobalExclusiveDeviceAccess GetOrCreate(
+            string portInstanceId,
+            int millisecondsTimeout,
+            CancellationToken? cancellationToken)
         {
-            for (bool retry = true; retry;)
+            if (cancellationToken?.IsCancellationRequested == true)
             {
-                retry = false;
+                return null;
+            }
 
-                var waitHandles = new List<WaitHandle>();
-                var mutex = new Mutex(false, $"{MutexBaseName}{connectionKey}");
-                waitHandles.Add(mutex);
+            // If the access lock has been created earlier (in previous statements) and has not yet been disposed,
+            // use that lock.
+            GlobalExclusiveDeviceAccess result = null;
 
-                CancellationTokenSource timeOutToken = null;
-                if (millisecondsTimeout > 0 && millisecondsTimeout != Timeout.Infinite)
+            lock (s_locks)
+            {
+                if (s_locks.TryGetValue(
+                    portInstanceId,
+                    out (AsyncLocal<GlobalExclusiveDeviceAccess>, Semaphore) instance))
                 {
-                    timeOutToken = new CancellationTokenSource(millisecondsTimeout);
-                    waitHandles.Add(timeOutToken.Token.WaitHandle);
+                    // Note that the result can still be null, in case the exclusive access was obtained by
+                    // code that is not a statement previously executed in the context of the current async-thread.
+                    result = instance.Item1?.Value;
                 }
+            }
 
-                if (cancellationToken.HasValue)
+            if (result is not null)
+            {
+                // Use the same lock as in Dispose
+                lock (result._mutex)
                 {
-                    waitHandles.Add(cancellationToken.Value.WaitHandle);
+                    if (result._lockCount == 0)
+                    {
+                        // It should already have been disposed of. Fix that for good measure.
+                        lock (s_locks)
+                        {
+                            s_locks.Remove(portInstanceId);
+                        }
+
+                        result = null;
+                    }
+                    else
+                    {
+                        result._lockCount++;
+                        return result;
+                    }
                 }
+            }
+
+            CancellationTokenSource timeOutToken = null;
+
+            if (millisecondsTimeout is >= 0 and not Timeout.Infinite)
```

Comment:
Does is make sense to check for millisecondsTimeout being 0? I don't think this is necessary... 🤔

Reply:
I was worried (I think) what happened if you would pass -2. In .NET that will result in an exception (from the constructor of the CancellationTokenSource). Removed the 0-comparison, so now this code will also throw an exception for a value of -2.